arangod/VocBase/vocbase.cpp

```diff
@@ -470,84 +461,54 @@ struct arangodb::VocBaseLogManager {
   Guarded<GuardedData> _guardedData;
 };
 
-/// @brief increase the reference counter for a database
-bool TRI_vocbase_t::use() {
-  auto expected = _refCount.load(std::memory_order_relaxed);
-  while (true) {
-    if ((expected & 1) != 0) {
-      // deleted bit is set
+bool TRI_vocbase_t::use() noexcept {
+  auto v = _refCount.load(std::memory_order_relaxed);
+  do {
+    if ((v & 1) != 0) {
       return false;
     }
-    // increase the reference counter by 2.
-    // this is because we use odd values to indicate that the database has been
-    // marked as deleted
-    auto updated = expected + 2;
-    TRI_ASSERT((updated & 1) == 0);
-    if (_refCount.compare_exchange_weak(expected, updated,
-                                        std::memory_order_release,
-                                        std::memory_order_relaxed)) {
-      // compare-exchange worked. we're done
-      return true;
-    }
-    // compare-exchange failed. try again!
-    expected = _refCount.load(std::memory_order_relaxed);
-  }
+  } while (!_refCount.compare_exchange_weak(v, v + 2, std::memory_order_acquire,
+                                            std::memory_order_relaxed));
+  return true;
 }
 
-void TRI_vocbase_t::forceUse() { _refCount += 2; }
+void TRI_vocbase_t::forceUse() noexcept {
+  _refCount.fetch_add(2, std::memory_order_relaxed);
+}
 
-/// @brief decrease the reference counter for a database
 void TRI_vocbase_t::release() noexcept {
-  // decrease the reference counter by 2.
-  // this is because we use odd values to indicate that the database has been
-  // marked as deleted
-  auto oldValue = _refCount.fetch_sub(2);
-  TRI_ASSERT(oldValue >= 2);
+  _refCount.fetch_sub(2, std::memory_order_release);
 }
 
-/// @brief returns whether the database can be dropped
-bool TRI_vocbase_t::isDangling() const {
-  if (isSystem()) {
-    return false;
-  }
-  auto refCount = _refCount.load();
-  // we are intentionally comparing with exactly 1 here, because a 1 means
-  // that noone else references the database but it has been marked as deleted
-  return (refCount == 1);
+bool TRI_vocbase_t::isDangling() const noexcept {
+  auto const v = _refCount.load(std::memory_order_acquire);
+  TRI_ASSERT((v & 1) == 0 || !isSystem());
+  return v == 1;
```

Comment:
I think this could be a behavior change and may be unsafe.
It someone manages to set the drop bit on the system database, this function will now return `true`, but previously it always returned `false`. It could have unintended side-effects for example in the DatabaseManagerThread, which will remove the database then if this function here returns true.
ofc it would be useful to not allow to set the drop bit for the system database in the first place, but this function here previously provided some extra security.

Reply:
No, please see assert in markAsDropped.
System database is never isDropped, so it's never isDangling

Reply:
I guess ASSERT than should be our choice. As system database should never have delete bit set.